[WIP] feat(UI): Implementa TTkbootstrap e refina telas de Menu/Refine (Fixes #3)

BrunoCGhisi opened the pull request:
closes #3 

## What this PR Addresses:
* [x] Initial development for Issue #3 (Improve Interface).
* [x] Implementation of the TTkbootstrap library for the interface theme.
* [x] Application of the new visual theme to the Menu screens.
* [x] **Additional Enhancement:** Addition of a new Menu Screen.
* [x] **Additional Enhancement:** Start of the Config Screen implementation.
* [x] **Additional Enhancement:** Code is open for music integration via the Pygame library.
* [x] **Additional Enhancement:** Responsiveness and minimum window size implemented.
* [x] **Additional Enhancement:** Menu centralization.

---

## Next Steps

### Branch 1.2-Refine-Interface-Details
* Switch the title bar icon.
* Switch the button icon.
* Refresh screen size automatically when N++ (Assuming 'N++' means incrementing a variable 'N').
* Place/Remove queen sound implementation.

### Branch 1.3-Improve-Config-Screen
* Config Screen (Will not be the root window, will spawn a separate frame/window) will contain:
    * Switch Bar for the music volume.
    * Option to turn off the place/remove queen sound.
    * Ability to alter the 'N' phase (only numbers that have already been passed through) (One to Nine phases).
    * Potential option to alter the TTkBootstrap Theme.

BrunoCGhisi committed:
Starting the componentization

BrunoCGhisi committed:
menu inicio

BrunoCGhisi committed:
add menu & music

BrunoCGhisi committed:
Configure responsiveness

- Add minsize in the root
- In the frame menu_screen:
** Remove pack geomatry
** Include grid geomatry

BrunoCGhisi committed:
Configure responsiviness

- Ajust font title
- Insertion 3 column to ajust resize

BrunoCGhisi committed:
Update README.md

Signed-off-by: Bruno Costa Ghisi <133908796+BrunoCGhisi@users.noreply.github.com>

BrunoCGhisi committed:
Update README.md

Signed-off-by: Bruno Costa Ghisi <133908796+BrunoCGhisi@users.noreply.github.com>

BrunoCGhisi committed:
Update README.md

Signed-off-by: Bruno Costa Ghisi <133908796+BrunoCGhisi@users.noreply.github.com>

BrunoCGhisi committed:
Menu Screen Improvements

~ Menu Screen Layout Inspired in the menu of "Balatro" video game ~
- Switch theme "Solar" to "Vapor".
- Adjust in minSize by the window & height size.
- Insertion ButtonContainer with subgrids.
- New button layout & font style.
- New button "Configuração" for spawn a Frame (To improve).

BrunoCGhisi committed:
No music, copyright exist.

- Remove the archives mp3.
- Comment all the music related codes.

iskigow commented on game.py:
```diff
@@ -1,99 +1,239 @@
-        self.start_time = None
+        self.menu_frame = ttk.Frame(root) 
+
+        # self.MUSIC_FILES = {
+        #     "menu": "music/BalatroMainTheme.mp3", 
+        #     "game": "music/BalatroBossTheme.mp3" 
+        # }
+        # self.current_music = None
+
+        # try:
+        #     pygame.mixer.init()
+        # except pygame.error as e:
+        #     print("Erro ao tocar a música: {e}. Música off")
+        #     self.MUSIC_FILES = {}
```
Maybe create an empty mp3 folder, so the player can put the mp3 they want to listen to.

BrunoCGhisi replied:
Do you mean something like this?

- Create an empty folder X.
- Create a folder Y with some default .mp3 music files.

This way, the user can add their own music to the empty folder.
If the folder X remains empty (meaning the user hasn’t added any music), the script will automatically use the default .mp3 files already included in the project folders.

iskigow replied:
only create an empty folder X, Only plays if have files there.

BrunoCGhisi replied:
Ok

BrunoCGhisi commented:
Hey @iskigow ,

I was thinking about something. When N becomes larger, the difficulty level of the phase actually becomes proportionally easier.

That happens because a bigger board means the player has more possible squares to place the queens, which consequently leads to more possible solutions.

I’ve also done some research and i'll be anexing one image that shows the "N x N values = Possible Solutions", of couse that in the context of our project, the number of possible solutions might differ a bit due to the COLORS AREA factor. However, the total number of solutions still tends to increase.

<img width="357" height="492" alt="image" src="https://github.com/user-attachments/assets/58b85e51-ffb8-400b-bc12-71a4db59b01c" />

My suggestion: to create a progressive difficulty system, we could start with N = 9, and every time the player gets the Win Message, the value of N decreases by 1.

What do you think?

BrunoCGhisi committed:
init working interface detils

- Icon root insertion
- Archives structure:
  Assets/
  |-- Images/queen-icon-ico.ico
  | -- Music

BrunoCGhisi committed:
Merge pull request #1 from BrunoCGhisi/1.2/fix/issue-3#ImproveInterface-Refine-Interface-Details

init working interface detils

iskigow commented:
> Hey @iskigow ,
> 
> I was thinking about something. When N becomes larger, the difficulty level of the phase actually becomes proportionally easier.
> 
> That happens because a bigger board means the player has more possible squares to place the queens, which consequently leads to more possible solutions.
> 
> I’ve also done some research and i'll be anexing one image that shows the "N x N values = Possible Solutions", of couse that in the context of our project, the number of possible solutions might differ a bit due to the COLORS AREA factor. However, the total number of solutions still tends to increase.
> <img alt="image" width="357" height="492" src="https://private-user-images.githubusercontent.com/133908796/504888055-58b85e51-ffb8-400b-bc12-71a4db59b01c.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NjEyNDY0MTcsIm5iZiI6MTc2MTI0NjExNywicGF0aCI6Ii8xMzM5MDg3OTYvNTA0ODg4MDU1LTU4Yjg1ZTUxLWZmYjgtNDAwYi1iYzEyLTcxYTRkYjU5YjAxYy5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUxMDIzJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MTAyM1QxOTAxNTdaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1hMjlkOTg1NzIwOGVhYzc2NzhhZTM3ZDY2NDM2NjE4ZTEyMThlM2ExMjgwOTlhMDM2NDU0ODI5NGY0M2FkMDUyJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.WHVXZwkbyucdNa28nERyV0Rke5EpmU9M1TvD5tEnW-0">
> 
> My suggestion: to create a progressive difficulty system, we could start with N = 9, and every time the player gets the Win Message, the value of N decreases by 1.
> 
> What do you think?

@BrunoCGhisi

Thank you for your insights and suggestion! But actually, the challenge of the problem becomes exponentially greater as the value of N increases, rather than easier.
Here are the main points:

- Complexity increases with N: While the number of solutions grows, so does the "search space" that the the player needs to explore to find them. The amount of computational work required to find the correct solution on a larger board is significantly greater.
- More options, more difficulty: A larger board with more squares can, paradoxically, lead to more dead ends, as the player has more positions that seem promising but don't lead to a final solution. This increases the need for "backtracking" (undoing moves) and makes the process of finding the solution far more complex.

So to create a progressive difficulty system, the best approach would be to increase the value of N with each player victory, as the challenge grows with the size of the board, by the way this is the problem related to issue #2

BrunoCGhisi commented:
> > Hey @iskigow ,
> > I was thinking about something. When N becomes larger, the difficulty level of the phase actually becomes proportionally easier.
> > That happens because a bigger board means the player has more possible squares to place the queens, which consequently leads to more possible solutions.
> > I’ve also done some research and i'll be anexing one image that shows the "N x N values = Possible Solutions", of couse that in the context of our project, the number of possible solutions might differ a bit due to the COLORS AREA factor. However, the total number of solutions still tends to increase.
> > <img alt="image" width="357" height="492" src="https://private-user-images.githubusercontent.com/133908796/504888055-58b85e51-ffb8-400b-bc12-71a4db59b01c.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NjEyNDY0MTcsIm5iZiI6MTc2MTI0NjExNywicGF0aCI6Ii8xMzM5MDg3OTYvNTA0ODg4MDU1LTU4Yjg1ZTUxLWZmYjgtNDAwYi1iYzEyLTcxYTRkYjU5YjAxYy5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUxMDIzJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MTAyM1QxOTAxNTdaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1hMjlkOTg1NzIwOGVhYzc2NzhhZTM3ZDY2NDM2NjE4ZTEyMThlM2ExMjgwOTlhMDM2NDU0ODI5NGY0M2FkMDUyJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.WHVXZwkbyucdNa28nERyV0Rke5EpmU9M1TvD5tEnW-0">
> > My suggestion: to create a progressive difficulty system, we could start with N = 9, and every time the player gets the Win Message, the value of N decreases by 1.
> > What do you think?
> 
> @BrunoCGhisi
> 
> Thank you for your insights and suggestion! But actually, the challenge of the problem becomes exponentially greater as the value of N increases, rather than easier. Here are the main points:
> 
> * Complexity increases with N: While the number of solutions grows, so does the "search space" that the the player needs to explore to find them. The amount of computational work required to find the correct solution on a larger board is significantly greater.
> * More options, more difficulty: A larger board with more squares can, paradoxically, lead to more dead ends, as the player has more positions that seem promising but don't lead to a final solution. This increases the need for "backtracking" (undoing moves) and makes the process of finding the solution far more complex.
> 
> So to create a progressive difficulty system, the best approach would be to increase the value of N with each player victory, as the challenge grows with the size of the board, by the way this is the problem related to issue #2

Thanks! I appreciate your confidence and liberty for suggestions of new ideas that you give to me.

Oh, now i understand your thinking line, your analyses measures the difficult level by more about the human behavior than the base numbers of solutions. Actually your analyses really makes more sense than my line of thinking.

When i finish the issue #3 i'lbe working in the #2  and #1 !

BrunoCGhisi commented:
Hi @iskigow!
In the last weeks i've been (and still am) working in other personal projects for improve some languages like Java and Js that I'm aiming to master. Because of this, i haven't made any new commits in the N-Queens. But now, i will comeback in the project with everything! Haha.

Ok, regarding the code, in the `readme`, i've implemented new features that i believe will improve the overall UX. All these items are being, or will be, implemented in the branch: **# 1.2-Refine-Interface-Details**.
<br>
<img width="822" height="227" alt="image" src="https://github.com/user-attachments/assets/e84ae445-0230-4961-b4d2-63ab2c4921eb" />

<br> 

> > ## Branch 1.2-Refine-Interface-Details 
>
> - Switch the title bar icon 
> - Switch the button icon
> - If has music file inserted in the game, will play the music of the player *@iskigow demand*
> - In game:
>     - Refresh screen size automaticaly when N++ 
>     - Increment in the interface the [numberPlaceQueens / numberTotalQueensToPlace]. For the UX
>     - When place a Queen in one color that already has a Queen, but is "placeble". Just switch the color acutual queen by the place that the player has inserted. For the UX.
>     - Place/Remove queen sound
>     - Centralize the board/canva

So, i appreciate your opinion and thoughts about it, what do you think? You have some additional requisition for these ideas?

Thanks!
